src/Microsoft.Diagnostics.Monitoring.RestServer/Controllers/DiagController.cs

```diff
@@ -59,25 +69,74 @@ public Task<ActionResult> GetDump(int? pid, [FromQuery]DumpType type = DumpType.
+                //Compression is done automatically by the response
+                //Chunking is done because the result has no content-length
+                return File(result.Value, "application/octet-stream", Invariant(dumpFileName));
             });
         }
 
         [HttpGet("cpuprofile/{pid?}")]
-        public Task<ActionResult> CpuProfile(int? pid, [FromQuery]int duration = 30)
+        public Task<ActionResult> CpuProfile(int? pid, [FromQuery]int durationSeconds = 30)
+        {
+            return InvokeService(async () =>
+            {
+                ValidateDuration(durationSeconds);
+                OperationResult<IStreamResult> result = await _diagnosticServices.StartCpuTrace(pid, durationSeconds, this.HttpContext.RequestAborted);
+                return new EventStreamResult(result.Value, "application/octet-stream", Invariant($"{Guid.NewGuid()}.nettrace"));
+            });
+        }
+
+        [HttpGet("trace/{pid?}")]
+        public Task<ActionResult> Trace(int? pid, [FromQuery]int durationSeconds = 30)
         {
             return InvokeService(async () =>
             {
-                OperationResult<Stream> result = await _diagnosticServices.StartCpuTrace(pid, duration, this.HttpContext.RequestAborted);
-                return CreateFileStreamResult(result, $"{Guid.NewGuid()}.nettrace");
+                ValidateDuration(durationSeconds);
+                OperationResult<IStreamResult> result = await _diagnosticServices.StartTrace(pid, durationSeconds, this.HttpContext.RequestAborted);
+                return new EventStreamResult(result.Value, "application/octet-stream", Invariant($"{Guid.NewGuid()}.nettrace"));
             });
         }
 
-        private FileStreamResult CreateFileStreamResult(OperationResult<Stream> result, FormattableString downloadName)
+        [HttpGet("logs/{pid?}")]
+        public ActionResult Logs(int? pid, [FromQuery] int durationSeconds = 30)
         {
-            //Compression is done automatically by the response
-            //Chunking is done because the result has no content-length
-            return File(result.Value, "application/octet-stream", Invariant(downloadName));
+            return InvokeService(() =>
+            {
+                ValidateDuration(durationSeconds);
+
+                return new OutputStreamResult(async (outputStream, token) =>
+                {
+                    await _diagnosticServices.StartLogs(outputStream, pid, durationSeconds, token);
+                }, "application/x-ndjson", Invariant($"{Guid.NewGuid()}.txt"));
+            });
+        }
+
+        private static void ValidateDuration(int durationSeconds)
+        {
+            if ((durationSeconds < 1) || (durationSeconds > MaxTraceSeconds))
+            {
+                throw new InvalidOperationException("Invalid duration");
+            }
+        }
+
+        private ActionResult InvokeService(Func<ActionResult> serviceCall)
```

Comment:
Unfortunately that won't work. Invoke Service can return different kinds of ActionResults. ActionResult<T> also doesn't inherit from ActionResult or IActionResult so it's challenging to merge those concepts.

Reply:
> Unfortunately that won't work. Invoke Service can return different kinds of ActionResults. 

I'm sorry I didn't follow? A concrete example perhaps where it doesn't work? (Feel free to merge without this, I trust you but I am curious to understand now)

> ActionResult also doesn't inherit from ActionResult …

Typo?

Reply:
BadRequest calls returns a concrete ActionResult, while the service call can return ActionResult or `ActionResult<T>`. There are some implicit conversions between these, but they are not part of the same hierarchy. 

I do believe there is a way to merge the implementations (relying on implicit and explicit conversions). I'll do some testing and send it up for review.

Reply:
Ah now I think I see your worry. You were concerned that given an arbitrary T the compiler can't guarantee that the return type of BadRequest is implicitly convertible to T. Because of the type hierarchy there is also no common type to use as a generic constraint. You could solve it with an explicit cast as below, but your code as-is is probably the best option. Sorry for the long detour.
````
        private T InvokeService<T>(Func<T> serviceCall)
        {
            try
            {
                return serviceCall();
            }
            catch (ArgumentException e)
            {
                return (T) BadRequest(FromException(e));
            }
            catch (DiagnosticsClientException e)
            {
                return (T) BadRequest(FromException(e));
            }
            catch (InvalidOperationException e)
            {
                return (T) BadRequest(FromException(e));
            }
        }
````
However if you worried that would allow people to invoke it with a bad T that will only be detected as a runtime InvalidCastException in the error case I'm fine with the original code too.

Reply:
even with a shared type you have to do strange things like:
(T)(ActionResult)BadRequest(). This is also very dangerous since T could be a concrete type rather than the shared base type.